### ISSUE TITLE
AttributeErrors are silently caught by trait_get()

### ISSUE THREAD

**Issue by nicolasap-dm** [enthought/traits#1747]
Given a class with a `_default` magic method that contains an `AttributeError`:

```python
from traits.api import HasStrictTraits, Float

class HasBrokenDefaults(HasStrictTraits):
    foo = Float()

    def _foo_default(self):
        self.missing_attr

has_broken_defaults = HasBrokenDefaults()
```

the following should cause the error to be raised, but instead just ignores the offending trait:

```python
has_broken_defaults.trait_get()
# returns: {}
```

Errors that are not AttributeErrors are correctly raised.

<details>

```python
from traits.api import HasStrictTraits, Float

class HasDifferentBrokenDefaults(HasStrictTraits):
    foo = Float()
    bar = Float()

    def _foo_default(self):
        self.missing_attr

    def _bar_default(self):
        1/0

has_different_broken_defaults = HasDifferentBrokenDefaults()

has_different_broken_defaults.trait_get()
# raises the ZeroDivisionError (as expected)

has_different_broken_defaults.trait_get(["foo"])
# does not raise. Instead returns: {}
```

</details>

**Comment by mdickinson** [comment]
> catch `AttributeError` in `_*_default` methods

Sorry, that wasn't clear. That is, catch `AttributeError` in the Traits library logic that executes those default methods, not within the code of the methods themselves.

**Comment by mdickinson** [comment]
> in the Traits library logic that executes those default methods

Specifically, [here](https://github.com/enthought/traits/blob/2b95e24aa77fcd70d190ac99190bfef2209f120a/traits/ctraits.c#L1823-L1824) and [here](https://github.com/enthought/traits/blob/2b95e24aa77fcd70d190ac99190bfef2209f120a/traits/ctraits.c#L1830).

**Comment by mdickinson** [comment]
Implementation note: ideally, if a warning is issued while handing the `AttributeError` and the current warnings settings are such that that warning is raised as an exception, then that exception would have the `AttributeError` as its `__context__` (or possibly even as its `__cause__`); this is what happens if you execute a `warnings.warn` call in an `except` handler in Python code. However, the C-API doesn't currently make it easy to do that exception chaining. That's unfortunate, because the information about where the `AttributeError` comes from would likely be useful to the developer.

Related: 
* https://stackoverflow.com/questions/51030659/modifying-or-reraising-python-error-in-c-api
* https://github.com/python/cpython/issues/67377

**Comment by mdickinson** [comment]
POC fix in #1762. With this fix, if I run the first script (with a `print(has_broken_defaults.trait_get())` appended), I get these results:

```python
(traits) mdickinson@mirzakhani traits % python issue1747.py
/Users/mdickinson/Enthought/ETS/traits/traits/has_traits.py:1413: UserWarning: default value resolution raised an AttributeError; this may cause Traits to behave in unexpected ways
  value = getattr(self, name, missing)
(traits) mdickinson@mirzakhani traits % python -Werror issue1747.py
Traceback (most recent call last):
  File "/Users/mdickinson/Enthought/ETS/traits/issue1747.py", line 7, in _foo_default
    self.missing_attr
AttributeError: 'HasBrokenDefaults' object has no attribute 'missing_attr'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/mdickinson/Enthought/ETS/traits/issue1747.py", line 11, in <module>
    has_broken_defaults.trait_get()
  File "/Users/mdickinson/Enthought/ETS/traits/traits/has_traits.py", line 1413, in trait_get
    value = getattr(self, name, missing)
UserWarning: default value resolution raised an AttributeError; this may cause Traits to behave in unexpected ways
```

The key thing to note is that the traceback in the second case includes the location of the `self.missing_attr` line, so that there's a reasonably path towards finding _where_ the troublesome `AttributeError` was raised.
